pkg/rpc/response/result/version.go

```diff
@@ -32,8 +66,121 @@ type (
+		MemoryPoolMaxTransactions   int             `json:"memorypoolmaxtransactions"`
+		ValidatorsCount             byte            `json:"validatorscount"`
+		InitialGasDistribution      json.RawMessage `json:"initialgasdistribution"`
+		StateRootInHeader           bool            `json:"staterootinheader,omitempty"`
 	}
 )
+
+// breakingVersion is a NeoGo revision that breaks older RPC clients compatibility with
+// newer RPC servers (https://github.com/nspcc-dev/neo-go/pull/2435).
+var breakingVersion = *semver.New("0.98.3-pre-89-g1c6afe40")
+
+// MarshalJSON implements the json marshaller interface.
+func (v *Version) MarshalJSON() ([]byte, error) {
+	aux := versionMarshallerAux{
+		Magic:     v.Magic,
+		TCPPort:   v.TCPPort,
+		WSPort:    v.WSPort,
+		Nonce:     v.Nonce,
+		UserAgent: v.UserAgent,
+		Protocol: protocolMarshallerAux{
+			AddressVersion:              v.Protocol.AddressVersion,
+			Network:                     v.Protocol.Network,
+			MillisecondsPerBlock:        v.Protocol.MillisecondsPerBlock,
+			MaxTraceableBlocks:          v.Protocol.MaxTraceableBlocks,
+			MaxValidUntilBlockIncrement: v.Protocol.MaxValidUntilBlockIncrement,
+			MaxTransactionsPerBlock:     v.Protocol.MaxTransactionsPerBlock,
+			MemoryPoolMaxTransactions:   v.Protocol.MemoryPoolMaxTransactions,
+			ValidatorsCount:             v.Protocol.ValidatorsCount,
+			InitialGasDistribution:      int64(v.Protocol.InitialGasDistribution),
+			StateRootInHeader:           v.Protocol.StateRootInHeader,
+		},
+		StateRootInHeader: v.StateRootInHeader,
+	}
+	return json.Marshal(aux)
+}
+
+// UnmarshalJSON implements the json unmarshaller interface.
+func (v *Version) UnmarshalJSON(data []byte) error {
+	var aux versionUnmarshallerAux
+	err := json.Unmarshal(data, &aux)
+	if err != nil {
+		return err
+	}
+	v.Magic = aux.Magic
+	v.TCPPort = aux.TCPPort
+	v.WSPort = aux.WSPort
+	v.Nonce = aux.Nonce
+	v.UserAgent = aux.UserAgent
+	v.Protocol.AddressVersion = aux.Protocol.AddressVersion
+	v.Protocol.Network = aux.Protocol.Network
+	v.Protocol.MillisecondsPerBlock = aux.Protocol.MillisecondsPerBlock
+	v.Protocol.MaxTraceableBlocks = aux.Protocol.MaxTraceableBlocks
+	v.Protocol.MaxValidUntilBlockIncrement = aux.Protocol.MaxValidUntilBlockIncrement
+	v.Protocol.MaxTransactionsPerBlock = aux.Protocol.MaxTransactionsPerBlock
+	v.Protocol.MemoryPoolMaxTransactions = aux.Protocol.MemoryPoolMaxTransactions
+	v.Protocol.ValidatorsCount = aux.Protocol.ValidatorsCount
+	v.Protocol.StateRootInHeader = aux.Protocol.StateRootInHeader
+	v.StateRootInHeader = aux.StateRootInHeader
+	if len(aux.Protocol.InitialGasDistribution) == 0 {
+		return nil
+	}
+
+	if strings.HasPrefix(v.UserAgent, config.UserAgentWrapper+config.UserAgentPrefix) {
+		ver, err := userAgentToVersion(v.UserAgent)
+		if err != nil {
+			// TODO: @roman-khimov, maybe we need to treat version parsing error as newer neo-go node and continue parsing using new rules?
+			return err
+		}
+		if ver.Compare(breakingVersion) < 0 {
```

Comment:
I'd do `ver.Compare("0.98.2")` <= 0 since we can't compare suffixes. It'll make any `0.98.3-pre-smth` new, but it's not a problem for us.

Reply:
Sure, fixed.